src/connectors/crowdstrike_devices.py

```diff
@@ -256,8 +308,258 @@ def ingest(table_name, options):
+    """
+    return authenticated_get_data(CROWDSTRIKE_VULN_ENTITY_URL, vuln_ids).get(
+        "resources", []
+    )
+
+
+async def insert_vuln_dets(table_name: str, dets: list) -> Coroutine[None, None, int]:
+    """
+    insert_vuln_dets inserts the vuln details into the database.
+    """
+    if len(dets) == 0:
+        return 0
+    numRows = db.insert(
+        table_name,
+        values=[{'raw': det, 'event_time': datetime.utcnow()} for det in dets],
+        select=db.derive_insert_select(LANDING_SPOTLIGHT_VULNS_TABLE_COLUMNS),
+        columns=db.derive_insert_columns(LANDING_SPOTLIGHT_VULNS_TABLE_COLUMNS),
+    )["number of rows inserted"]
+
+    log.info(f"[UPLOAD]: Inserting {numRows} rows in vuln details")
+    return numRows
+
+
+async def get_and_insert_rem_dets(vuln_dets: list) -> Coroutine[None, None, int]:
+    """
+    get_and_insert_rem_dets dedupes and remediation IDs and then fetches their
+    details from the API. The results are then inserted into the database.
+    """
+    dets = await get_uncached_rem_dets(vuln_dets)
+    return await insert_rem_dets(dets)
+
+
+async def pull_spotlight_vuln_details(
+    table_name: str, vuln_ids: Tuple[str, str]
+) -> Coroutine[None, None, list]:
+    """
+    pull_spotlight_vuln_details gets all the vuln detail information, inserts it
+    into the database and retrieves any remediation IDs and pulls those into Snowflake
+    as well.
+    """
+    dets: list = await get_vuln_dets(vuln_ids)
+    return await asyncio.gather(
+        insert_vuln_dets(table_name, dets),
+        get_and_insert_rem_dets(dets),
+    )
+
+
+def authenticated_get_data_builder(options: Dict[str, str]):
+    token = get_token_basic(
+        options.get('client_id', ""), options.get('client_secret', "")
+    )
+
+    def request(url: str, params: Dict[str, str]):
+        nonlocal token
+        while True:
+            try:
+                return get_data(token, url, params)
+            except requests.HTTPError as http_err:
+                log.info(http_err)
+                if http_err.errno == 401 | http_err.errno == 403:
+                    log.info("Token expired, getting new token...")
+                    token = get_token_basic(
+                        options.get('client_id', ""), options.get('client_secret', "")
+                    )
+                    log.info(f"Retying request: {url} {params}")
+
+    return request
+
+
+def authenticated_get_data(url: str, params: Dict[str, str]):
+    log.info("NOT IMPLEMENTED")
+
+
+def paginate_api(
+    params: Dict[str, str],
+    url: str,
+    max_page_entries: int = 400,
+):
+    """
+    paginate_api returns a generator that yields page entries from the Crowdstrike
+    Spotlight API. By default, page sizes are only 400 entries long because that maps
+    to the maximim number of vuln IDs that can be specified in the vuln details API, but
+    this is configurable.
```

Comment:
```suggestion
    paginate_api returns a generator that yields page entries from the specified Crowdstrike
    API. By default, page sizes are only 400 entries long because that maps
    to the maximim number of vuln IDs that can be specified in the vuln details API, but
    this is configurable.
```